File: src/features.py
"""Pre-game features from prior games only (shifted rolling stats)."""

from __future__ import annotations

import pandas as pd


def _team_long_table(games: pd.DataFrame) -> pd.DataFrame:
    """Each team-game as one row for computing rolling team form."""
    home = games.assign(
        TEAM_ID=games["HOME_TEAM_ID"],
        OPP_ID=games["AWAY_TEAM_ID"],
        IS_HOME=1,
        WIN=games["HOME_WIN"],
        PF=games["HOME_PTS"],
        PA=games["AWAY_PTS"],
    )[["GAME_ID", "GAME_DATE", "TEAM_ID", "OPP_ID", "IS_HOME", "WIN", "PF", "PA"]]

    away = games.assign(
        TEAM_ID=games["AWAY_TEAM_ID"],
        OPP_ID=games["HOME_TEAM_ID"],
        IS_HOME=0,
        WIN=1 - games["HOME_WIN"],
        PF=games["AWAY_PTS"],
        PA=games["HOME_PTS"],
    )[["GAME_ID", "GAME_DATE", "TEAM_ID", "OPP_ID", "IS_HOME", "WIN", "PF", "PA"]]

    long_df = pd.concat([home, away], ignore_index=True)
    long_df = long_df.sort_values(["TEAM_ID", "GAME_DATE", "GAME_ID"]).reset_index(drop=True)
    return long_df


def add_shifted_roll_features(games: pd.DataFrame, window: int = 10, min_periods: int = 3) -> pd.DataFrame:
    long_df = _team_long_table(games)

    def _roll(g: pd.DataFrame) -> pd.DataFrame:
        g = g.copy()
        w = g["WIN"].shift(1)
        margin = (g["PF"] - g["PA"]).shift(1)
        g["WIN_RATE_L10"] = w.rolling(window, min_periods=min_periods).mean()
        g["MARGIN_L10"] = margin.rolling(window, min_periods=min_periods).mean()
        g["REST_DAYS"] = g["GAME_DATE"].diff().dt.days
        return g

    long_df = long_df.groupby("TEAM_ID", group_keys=False).apply(_roll)

    home_side = long_df[long_df["IS_HOME"] == 1][
        ["GAME_ID", "WIN_RATE_L10", "MARGIN_L10", "REST_DAYS"]
    ].rename(
        columns={
            "WIN_RATE_L10": "HOME_WIN_RATE_L10",
            "MARGIN_L10": "HOME_MARGIN_L10",
            "REST_DAYS": "HOME_REST_DAYS",
        }
    )
    away_side = long_df[long_df["IS_HOME"] == 0][
        ["GAME_ID", "WIN_RATE_L10", "MARGIN_L10", "REST_DAYS"]
    ].rename(
        columns={
            "WIN_RATE_L10": "AWAY_WIN_RATE_L10",
            "MARGIN_L10": "AWAY_MARGIN_L10",
            "REST_DAYS": "AWAY_REST_DAYS",
        }
    )

    out = games.merge(home_side, on="GAME_ID", how="left").merge(away_side, on="GAME_ID", how="left")
    return out


def feature_matrix(df: pd.DataFrame) -> tuple[pd.DataFrame, list[str]]:
    feature_cols = [
        "HOME_WIN_RATE_L10",
        "HOME_MARGIN_L10",
        "HOME_REST_DAYS",
        "AWAY_WIN_RATE_L10",
        "AWAY_MARGIN_L10",
        "AWAY_REST_DAYS",
    ]
    X = df[feature_cols].fillna(0.0)
    return X, feature_cols

File: web/src/lib/features.ts
import type { CompletedGame } from "./types";

const FEATURE_COLS = [
  "HOME_WIN_RATE_L10",
  "HOME_MARGIN_L10",
  "HOME_REST_DAYS",
  "AWAY_WIN_RATE_L10",
  "AWAY_MARGIN_L10",
  "AWAY_REST_DAYS",
] as const;

interface TeamLongRow {
  GAME_ID: string;
  GAME_DATE: string;
  TEAM_ID: number;
  IS_HOME: number;
  WIN: number;
  PF: number;
  PA: number;
}

function parseDate(s: string): number {
  return new Date(`${s}T12:00:00`).getTime();
}

function teamLongTable(games: CompletedGame[]): TeamLongRow[] {
  const rows: TeamLongRow[] = [];
  for (const g of games) {
    rows.push({
      GAME_ID: g.GAME_ID,
      GAME_DATE: g.GAME_DATE,
      TEAM_ID: g.HOME_TEAM_ID,
      IS_HOME: 1,
      WIN: g.HOME_WIN,
      PF: g.HOME_PTS,
      PA: g.AWAY_PTS,
    });
    rows.push({
      GAME_ID: g.GAME_ID,
      GAME_DATE: g.GAME_DATE,
      TEAM_ID: g.AWAY_TEAM_ID,
      IS_HOME: 0,
      WIN: 1 - g.HOME_WIN,
      PF: g.AWAY_PTS,
      PA: g.HOME_PTS,
    });
  }
  rows.sort((a, b) => {
    if (a.TEAM_ID !== b.TEAM_ID) return a.TEAM_ID - b.TEAM_ID;
    const da = parseDate(a.GAME_DATE);
    const db = parseDate(b.GAME_DATE);
    if (da !== db) return da - db;
    return a.GAME_ID.localeCompare(b.GAME_ID);
  });
  return rows;
}

function rollingMean(values: (number | null)[], window: number, minPeriods: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < values.length; i++) {
    const slice = values.slice(Math.max(0, i - window + 1), i + 1).filter((v): v is number => v !== null);
    if (slice.length < minPeriods) {
      out.push(NaN);
    } else {
      out.push(slice.reduce((a, b) => a + b, 0) / slice.length);
    }
  }
  return out;
}

function groupRoll(long: TeamLongRow[], window = 10, minPeriods = 3): Map<string, Record<string, number>> {
  const byGame = new Map<string, Record<string, number>>();
  const groups = new Map<number, TeamLongRow[]>();
  for (const row of long) {
    const g = groups.get(row.TEAM_ID) ?? [];
    g.push(row);
    groups.set(row.TEAM_ID, g);
  }

  for (const teamRows of groups.values()) {
    const wins = teamRows.map((r) => r.WIN);
    const shiftedWins: (number | null)[] = [null, ...wins.slice(0, -1)];
    const margins: (number | null)[] = [null, ...teamRows.slice(0, -1).map((r) => r.PF - r.PA)];

    const winRate = rollingMean(shiftedWins, window, minPeriods);
    const margin = rollingMean(margins, window, minPeriods);

    for (let i = 0; i < teamRows.length; i++) {
      const row = teamRows[i];
      const restDays = i === 0 ? NaN : (parseDate(row.GAME_DATE) - parseDate(teamRows[i - 1].GAME_DATE)) / 86400000;

      const prefix = row.IS_HOME === 1 ? "HOME" : "AWAY";
      const existing = byGame.get(row.GAME_ID) ?? {};
      existing[`${prefix}_WIN_RATE_L10`] = Number.isNaN(winRate[i]) ? 0 : winRate[i];
      existing[`${prefix}_MARGIN_L10`] = Number.isNaN(margin[i]) ? 0 : margin[i];
      existing[`${prefix}_REST_DAYS`] = Number.isNaN(restDays) ? 0 : restDays;
      byGame.set(row.GAME_ID, existing);
    }
  }
  return byGame;
}

export function featureVectorForGame(games: CompletedGame[], gameId: string): number[] {
  const long = teamLongTable(games);
  const feats = groupRoll(long);
  const row = feats.get(gameId);
  if (!row) throw new Error("Features missing for game");
  return FEATURE_COLS.map((c) => row[c] ?? 0);
}

export function featureColumns(): readonly string[] {
  return FEATURE_COLS;
}
